chalk-engine/src/logic.rs

```diff
@@ -679,16 +678,29 @@ impl<C: Context> Forest<C> {
         let (table, subst, constraints, delayed_subgoals) =
             context.instantiate_answer_subst(num_universes, &answer.subst);
 
+        // FIXME: really, these shouldn't be added to the delayed_subgoals at all
+        // however, because we can't compare the table goal until `canonicalize_answer_subst`
+        // is called in `pursue_answer`, this will require a bit of refactoring work
```

Comment:
I don't really understand this comment -- oh, are you saying that "really there shouldn't be anything in `delayed_subgoals` that is equal to `table_goal`? If so, I'm not sure I agree, I think that can arise after unification, but you are correct that detecting that equality cannot reliably be done with `==`. (But you don't need unification, we could make a method that does it.) Anyway, if that's what you mean, I agree, but we should probably file an issue and/or write a better comment.

Reply:
Ok, so there's a couple things going on here, and I'll try to explain it succinctly since I'm traveling today (so on mobile at least for the minute).

So first, it's maybe not obvious from the PR (and maybe not even *wrong*), but something I found when looking through `CHALK_DEBUG` output. But in `apply_answer_subst`, we can't *properly* check if the delayed subgoals are equal to the table goal, because the delayed subgoals get unique inference variables. (Perhaps this would be easier/quicker to check using the variant3 test and adding dbg statements there.)

Second, I would REALLY prefer to have these be checked *before* they are added to the answer. But, because of 1,  we have to wait until they are both canonicalized, which happens in pursue_answer. At that point it is *too late* because that doesn't know about the subgoal. That was mostly the point of this FIXME: we can be smarter here I think, but it requires a bit of refactoring.

Third, there are essentially two places that do this check. First is in `apply_answer_subst` and second is here. They aren't the same: removing this causes the variant3 test to fail. Removing the other causes...many of the conductive tests to fail (and a never-ending test for one that I don't remember off the top of my head). I'm not convinced that there couldn't be just one check in the right place.

So, overall. I'm a bit skeptical. Between both checks, everything works. But as I said, I don't know if it's perfectly correct and I definitely think there's room for improvement. Whether that needs to be done in this PR or can be done separately idk. 

Nonetheless, I should have some time soon to, at the very least, cleanup this FIXME and a couple other things. (Like the extra apply_binders function that was added but isn't used anymore). If we decide to land the PR as is, I'll file a followup issue with more detailed info.

Reply:
OK, thanks for elaborating. Personally I'm inclined to convert this comment into an issue and land the PR, since I think it's clearly 'progress', but I definitely agree we can refactor this part of it. I'd be happy to take some time to chat it out.